### ISSUE TITLE
Compute Loss on Accumulated Batches

### ISSUE THREAD

**Issue by fschlatt** [PyTorchLightning/pytorch-lightning#992]
## 🚀 Feature
Make it possible to accumulate the outputs and targets from the training step for the number of grad batches that should be accumulated. Then compute the loss over the accumulated outputs and targets.

### Motivation
The accumulate grad batches option is great for people who only have access to a GPU with limited memory. However, some loss functions are not batch independent, i.e. the loss of one sample depends on the other samples within a batch. For example the F1 score or Matthew's correlation coefficient can be modified so they can be directly applied as loss functions. In these cases, it would be great to be able to accumulate the outputs and compute the loss afterwards, similarly to how it works in `validation_end` and `test_end`, except before the backwards step.

### Pitch

Add an `accumulate` function that takes the accumulated output of the `training_step`. The default behavior would be to output the sum over all keys in the dict and to normalize by the number of accumulated grad batches. This can then be overwritten though to compute the loss over the entire accumulated batch. This would not break the current API but accommodate batch dependent loss functions.

**Comment by Borda** [comment]
@fschlatt I agree that accumulator is very important, may you check [GradientAccumulationScheduler](https://pytorch-lightning.readthedocs.io/en/latest/pytorch_lightning.callbacks.html#pytorch_lightning.callbacks.GradientAccumulationScheduler)?

**Comment by fschlatt** [comment]
Never mind, my issue can be solved using [training_step_end](https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html#training-step-end) :)

**Comment by karasikov** [comment]
Hi @fschlatt, I'm trying to solve exactly the same problem. I have a contrastive loss that depends on all elements of the batch and, hence, has to be computed on the predictions accumulated from the entire batch.

Following this discussion, I'd think you simply compute the loss in `training_step_end` instead of `training_step`, however, it doesn't work if I do that. `training_step` is required to return a dict that already contains the loss.

How did you manage to solve this issue 4 years ago?
Thanks!

**Comment by fschlatt** [comment]
The solution has changed since 4 years ago, now the best way is documented here https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#train-epoch-level-operations

**Comment by karasikov** [comment]
Hi @fschlatt , thanks for the quick response!
I checked the docs but it's still unclear to me how to compute the loss on the entire batch.

The docs say that the loss should be computed in `training_step`, which is called after every step.
However, I need to compute it for the entire "accumulated" batch. Say, if I set `accumulate_grad_batches` to 32, I need to compute the loss once for all the outputs aggregated from all those 32 steps. What is the currently recommended way to do that?

Maybe this has a very simple solution, but I'm somehow missing it.
Thanks, really appreciate your help!

**Comment by fschlatt** [comment]
Sorry, my response was a bit too hasty. I haven't tested this, but I believe you are free to output anything you like from the `training_step`. For example, you can output the model predictions and desired targets in a dictionary.

 You can then aggregate these outputs in `train_batch_end` (which is where the loss is aggregated for `accumulate_grad_batches`) and compute the loss there
